lib/voice/VoiceConnection.js

```diff
@@ -464,6 +559,52 @@ class VoiceConnection extends EventEmitter {
         });
     }
 
+    #reinitDaveSession(userId, channelId) {
+        if(this.daveProtocolVersion > 0) {
+            if(this.daveSession) {
+                this.daveSession.reinit(this.daveProtocolVersion, userId, channelId);
+                this.emit("debug", `DAVE session reinitialized for protocol version ${this.daveProtocolVersion}`);
+            } else {
+                this.daveSession = new Davey.DAVESession(this.daveProtocolVersion, userId, channelId);
+                this.emit("debug", `DAVE session initialized for protocol version ${this.daveProtocolVersion}`);
+            }
+            this.sendWSBinary(VoiceOPCodes.MLS_KEY_PACKAGE, this.daveSession.getSerializedKeyPackage());
+        } else {
+            const daveSessionExisted = !!this.daveSession;
+            this.daveSession?.reset();
+            this.daveSession = null;
+            if(daveSessionExisted) {
+                this.emit("debug", "DAVE session reset");
+            }
+        }
+    }
+
+    #recoverFromInvalidCommit(transitionId, userId, channelId) {
+        this.sendWS(VoiceOPCodes.MLS_INVALID_COMMIT_WELCOME, {transition_id: transitionId});
+        this.#reinitDaveSession(userId, channelId);
+    }
+
+    #executePendingTransition(transitionId) {
+        if(!this.#davePendingTransition) {
+            return this.emit("warn", `Recieved execute transition, but we dont have a pending transition (${transitionId})`);
+        }
+        let transitioned = false;
+        if(transitionId !== this.#davePendingTransition.transition_id) {
+            this.emit("warn", `Recieved execute transition for an unexpected transition id (expected: ${this.#davePendingTransition.transition_id}, actual: ${transitionId})`);
+        } else {
+            const oldVersion = this.daveProtocolVersion;
+            this.daveProtocolVersion = this.#davePendingTransition.protocol_version;
+            transitioned = true;
+            this.emit("debug", `DAVE transition executed (v${oldVersion} -> v${this.daveProtocolVersion}, id: ${transitionId})`);
+        }
+        this.#davePendingTransition = null;
+        return transitioned;
+    }
+
+    #daveReady() {
```

Comment:
consider turning this to a getter

```suggestion
    get #daveReady() {
```